frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py

```diff
@@ -0,0 +1,851 @@
+    outvals = [cudaq_spin(idx, cuda_name)]
+    outvariables = [ctx.new_variable(AbsCudaSpinOperator())]
+
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, outvals)
+
+
+def change_hamiltonian(ctx, eqn):
+    """Change catalyst hamiltonian to an equivalent expression in CUDA."""
+    assert eqn.primitive == hamiltonian_p
+
+    invals = safe_map(ctx.read, eqn.invars)
+    coeffs = invals[0]
+    terms = invals[1:]
+
+    hamiltonian = reduce(operator.add, map(operator.mul, coeffs, terms))
+
+    outvariables = [ctx.new_variable(AbsCudaSpinOperator())]
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, [hamiltonian])
+
+
+def change_adjoint(ctx, eqn):
+    """Change Catalyst adjoint to an equivalent expression in CUDA."""
+    assert eqn.primitive == adjoint_p
+
+    # This is the quantum register.
+    invals = safe_map(ctx.read, eqn.invars)
+    register = invals[0]
+
+    # We might need this for pytrees.
+    _args_tree = eqn.params["args_tree"]
+    nested_jaxpr = eqn.params["jaxpr"]
+
+    # So, first we need to make a new kernel.
+    # And we need to pass a register type as an argument
+    # we are working on as an argument.
+    # cudaq.qreg is essentially an abstract type in cudaq
+    kernel_to_adjoint, abstract_qreg = cudaq_make_kernel(cudaq.qreg)
+
+    # We need a new interpreter with a new kernel
+    # And the parameter is abstract_qreg.
+    interpreter = InterpreterContext(
+        nested_jaxpr.jaxpr,
+        nested_jaxpr.literals,
+        abstract_qreg,
+        kernel=kernel_to_adjoint,
+    )
+    # retval would be the abstract_qreg we passed as an argument here.
+    # We always pass the qreg and return the qreg.
+    # TODO: Do we support returning any other values?
+    _retval = interpret_impl(interpreter, nested_jaxpr.jaxpr)
+    cudaq_adjoint(ctx.kernel, kernel_to_adjoint, register)
+
+    # Now that we have captured all of the adjoint...
+    # Let's actually create an operation with it.
+    safe_map(ctx.write, eqn.outvars, [register])
+
+
+def ignore_impl(_ctx, _eqn):
+    """No-op"""
+
+
+def unimplemented_impl(_ctx, eqn):
+    """Raise an error."""
+    msg = f"{eqn.primitive} is not yet implemented in Catalyst's CUDA-Quantum support."
+    raise NotImplementedError(msg)
+
+
+def default_impl(ctx, eqn):
+    """Default implementation for all other non-catalyst primitives. I.e., other JAX primitives."""
+    # This little scope was based on eval_jaxpr's implmentation:
+    #    https://github.com/google/jax/blob/16636f9c97414d0c5195c6fd47227756d4754095/jax/_src/core.py#L507-L518
+    subfuns, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    ans = eqn.primitive.bind(*subfuns, *map(ctx.read, eqn.invars), **bind_params)
+    if eqn.primitive.multiple_results:  # pragma: nocover
+        safe_map(ctx.write, eqn.outvars, ans)
+    else:
+        ctx.write(eqn.outvars[0], ans)
+
+
+INST_IMPL = {
```

Comment:
could you move this to the top of this file?

Reply:
Is it necessary? The use is immediately below. I don't like defining things that are too far away from where they are used. It makes reading code more difficult in my opinion. I originally planned on leaving it inlined in `interpret_impl` as it is only used there, but making a dictionary every time the function is called is not ideal.